osquery/tables/system/linux/process_open_pipes.cpp

```diff
@@ -0,0 +1,193 @@
+/**
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed in accordance with the terms specified in
+ *  the LICENSE file found in the root directory of this source tree.
+ */
+
+#include <iostream>
+#include <osquery/core.h>
+#include <osquery/filesystem/filesystem.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+#include <regex>
+
+namespace osquery {
+namespace tables {
+
+struct pipe_info {
+  pid_t pid;
+  int fd;
+  std::string mode;
+  ino_t inode;
+  std::string type;
+  pipe_info(pid_t pid, int fd, std::string mode, ino_t inode, std::string type)
+      : pid(pid), fd(fd), mode(mode), inode(inode), type(type) {}
+};
+
+using PidToPipesMap = std::map<pid_t, std::vector<pipe_info*>>;
+using InodeToPipesMap = std::map<ino_t, std::vector<pipe_info*>>;
+
+bool isSamePipe(const pipe_info* p1, const pipe_info* p2) {
+  return (p1->pid == p2->pid && p1->inode == p2->inode && p1->fd == p2->fd &&
+          p1->mode == p2->mode);
+}
+
+bool isReaderWriterPair(const pipe_info* p1, const pipe_info* p2) {
+  // Assumption: p1 and p2 have same inode
+  return ((p1->mode != p2->mode) || (p1->mode == "rw"));
+}
+
+bool isUnconnectedPipe(ino_t inode, const InodeToPipesMap& pipe_partners) {
+  if (pipe_partners.at(inode).size() == 1)
+    return true;
+  return false;
+}
+
+int parseInode(const std::string& pipe_str) {
+  std::smatch match;
+  if (std::regex_search(pipe_str, match, std::regex("\\d+")))
+    return std::stoi(match[0]);
+  else
+    return 0;
+}
+
+std::string getMode(const std::string& pid, const std::string& fd) {
+  std::string mode = "";
+  struct stat file_stat;
+  std::string filename = std::string("/proc/") + pid + "/fd/" + fd;
+  if (lstat(filename.c_str(), &file_stat))
+    return "-";
+  if (file_stat.st_mode & S_IRUSR || file_stat.st_mode & S_IRGRP ||
+      file_stat.st_mode & S_IROTH)
+    mode += "r";
+  if (file_stat.st_mode & S_IWUSR || file_stat.st_mode & S_IWGRP ||
+      file_stat.st_mode & S_IWOTH)
+    mode += "w";
+  return mode;
+}
+
+std::unique_ptr<pipe_info> createPipeInfoStruct(
+    pid_t pid, int fd, std::string mode, ino_t inode, std::string type) {
+  auto ps = std::make_unique<pipe_info>(pid, fd, mode, inode, type);
+  if (!ps)
+    LOG(ERROR) << "Error creating pipe_info for pid:" << pid << "\n";
```

Comment:
Space between `:` and `pid`, and glog will add a newline for you, you can remove the `\n`.